src/_pytest/tmpdir.py

```diff
@@ -214,3 +229,25 @@ def tmp_path(request: FixtureRequest, tmp_path_factory: TempPathFactory) -> Path
     """
 
     return _mk_tmp(request, tmp_path_factory)
+
+
+def pytest_sessionfinish(session, exitstatus: Union[int, ExitCode]):
+    """After each session, remove base directory if all the tests passed,
+    the policy is "failed", and the basetemp is not specified by a user.
+    """
+    tmp_path_factory: TempPathFactory = session.config._tmp_path_factory
+    if tmp_path_factory._basetemp is None:
+        return
+    policy = tmp_path_factory._retention_policy
```

Comment:
I was under the impression that the `failed` policy would remove all passed tests, even if some tests failed.

Here it will remove all directories only if all tests pass.

Or is that handled somewhere else that I am missing?

Reply:
> Here it will remove all directories only if all tests pass.

This is actually how I implemented it. And I added some implementation to match the behavior you mentioned.
[0abbf8c](https://github.com/pytest-dev/pytest/pull/10442/commits/0abbf8c0565562779700892a704adf7001a7352a)